tensorflow/python/ops/image_ops_impl.py

```diff
@@ -2531,7 +2531,7 @@ def convert_image_dtype(image, dtype, saturate=False, name=None):
         return math_ops.multiply(cast, scale, name=name)
       else:
         # Converting from float: first scale, then cast
-        scale = dtype.max + 0.5  # avoid rounding problems in the cast
+        scale = 1./(dtype.max + 0.5)  # avoid rounding problems in the cast
```

Comment:
This is not correct.

The case covered here is the input being floating-point and the output being an integer type.

Let's say you start with an image in the range [0, 1.0], then you are converting to uint8.  You want 1.0 to map to 255.  That's why the scale factor is by 255.5 - so that 255.5 * 1.0, rounded, is 255.

Reply:
No, we cannot do validation via numpy.  Tensors may not even be evaluated yet in graph mode to be able to access numpy values.

I don't think you should automatically scale the image either.

If `saturate` is `True`, it will do a saturated cast, which would clip anything outside of the range to the maximum value.  I think this is already the best that could be done.  Otherwise, if `saturate` is `False`, we do a blind cast which can result in integer overflows that cause values to wrap around.  For signed integers, this is actually "undefined behavior", which is why we're seeing differences between devices - essentially the results can't be trusted in either case.  The fact that we _already_ have a `saturate` option implies that some values may already be expected to be out of range, and explicitly setting `saturate=True` is the proper solution.